pkg/controllers/hostedcluster/controller/controller.go

```diff
@@ -0,0 +1,215 @@
+	}, &hostedClusterKubeconfigSecret); err != nil {
+		log.Error(err, "unable to get kubeconfig secret for hostedcluster", "hostedcluster", req.NamespacedName)
+		return ctrl.Result{}, err
+	}
+
+	// create managedcluster namespace for hostedcluster
+	if _, createErr := r.ControlplaneKubeClient.CoreV1().Namespaces().Create(ctx, &corev1.Namespace{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: req.NamespacedName.Name,
+		},
+	}, metav1.CreateOptions{}); createErr != nil {
+		if !errors.IsAlreadyExists(createErr) {
+			log.Error(createErr, "unable to create managedcluster namespace for hostedcluster", "hostedcluster", req.NamespacedName)
+			return ctrl.Result{}, createErr
+		}
+	}
+
+	// create managedcluster-kubeconfig for hostedcluster
+	if _, createErr := r.ControlplaneKubeClient.CoreV1().Secrets(req.NamespacedName.Name).Create(ctx, &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      helpers.ManagedClusterKubeConfig,
+			Namespace: req.NamespacedName.Name,
+		},
+		Data: hostedClusterKubeconfigSecret.Data,
+		Type: corev1.SecretTypeOpaque,
+	}, metav1.CreateOptions{}); createErr != nil {
+		if !errors.IsAlreadyExists(createErr) {
+			log.Error(createErr, "unable to create managedcluster-kubeconfig for hostedcluster", "hostedcluster", req.NamespacedName)
+			return ctrl.Result{}, createErr
+		}
+	}
+
+	// create klusterlet for hostedcluster
+	if _, createErr := r.ControlplaneOperatorClient.OperatorV1().Klusterlets().Create(ctx, &operatorapiv1.Klusterlet{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: req.NamespacedName.Name,
+			Annotations: map[string]string{
+				helpers.KlusterletOwnerAnnotation: "hostedcluster-controller",
+			},
+		},
+		Spec: operatorapiv1.KlusterletSpec{
+			DeployOption: operatorapiv1.KlusterletDeployOption{
+				Mode: operatorapiv1.InstallModeHosted,
+			},
+		},
+	}, metav1.CreateOptions{}); createErr != nil {
+		if !errors.IsAlreadyExists(createErr) {
+			log.Error(createErr, "unable to create klusterlet for hostedcluster", "hostedcluster", req.NamespacedName)
+			return ctrl.Result{}, createErr
+		}
+	}
+
+	return ctrl.Result{}, nil
+}
+
+func (r *HostedClusterReconciler) pruneManagedClusterResources(ctx context.Context, log logr.Logger, hostedCluster *hyperv1beta1.HostedCluster) error {
+	// remove klusterlet for hosted cluster
+	if err := r.ControlplaneOperatorClient.OperatorV1().Klusterlets().Delete(ctx, hostedCluster.GetName(), metav1.DeleteOptions{}); err != nil {
+		if !errors.IsNotFound(err) {
+			log.Error(err, "unable to delete klusterlet for hostedcluster", "hostedcluster", fmt.Sprintf("%s/%s", hostedCluster.GetNamespace(), hostedCluster.GetName()))
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (r *HostedClusterReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	hostedClusterPredicate := predicate.Funcs{
+		CreateFunc: func(e event.CreateEvent) bool {
+			if e.Object.GetAnnotations()[hostedClusterImportAnnotation] == "false" {
```

Comment:
who add this annotation? it seems that nobody to add this

Reply:
this is by design, see https://issues.redhat.com/browse/ACM-5588